### ISSUE TITLE
A parameter cannot be found that matches parameter name 'Chord'.

### ISSUE THREAD

**Issue by DesuNinshu** [PowerShell/vscode-powershell#4409]
### Prerequisites

- [X] I have written a descriptive issue title.
- [X] I have searched all [_open and closed_ issues](https://github.com/PowerShell/vscode-powershell/issues?q=is%3Aissue) to ensure it has not already been reported.
- [x] I have read the [troubleshooting](https://github.com/PowerShell/vscode-powershell/blob/main/docs/troubleshooting.md) guide.
- [x] I am sure this issue is with the _extension itself_ and does not reproduce in a standalone [PowerShell](https://github.com/PowerShell/PowerShell/issues/new/choose) instance.
- [x] I have verified that I am using the latest version of Visual Studio Code and the PowerShell extension.
- [x] If this is a security issue, I have read the [security issue reporting guidance](https://github.com/PowerShell/vscode-powershell/blob/main/SECURITY.md).

### Summary

Hello,

I discovered this bug when I ran the command below. It's worth noting that the command still works but it always puts out the error, even when I tried changing my error action preference.

I tried running a number of other commands for the command parameter and I put it into a script file and got the same result.
I don't have the Preview PowerShell extension enabled. I was running the command in my integrated terminal with the PowerShell extension `v2023.1.0`.

I tried running Get-Error and it wouldn't return the details of the error. I also tried stepping through the script I put it into, and it also couldn't catch it.

I am still somewhat new to PowerShell; I may not have given enough detail here.
Let me know if there are additional details I can provide.

## Edit:
Actually, I think I figured it out.

The `Get-PSReadLineKeyHandler` function does not have the chord parameter in the PowerShell 5.1 version of the function.
[Get-PSReadLineKeyHandler PowerShell-5.1](https://learn.microsoft.com/en-us/powershell/module/psreadline/get-psreadlinekeyhandler?view=powershell-5.1)

```PowerShell
Get-PSReadLineKeyHandler
   [-Bound]
   [-Unbound]
   [<CommonParameters>]
```

But it does in the 7.3 version. 
[Get-PSReadLineKeyHandler PowerShell-7.3](https://learn.microsoft.com/en-us/powershell/module/psreadline/get-psreadlinekeyhandler?view=powershell-7.3)

```PowerShell
Get-PSReadLineKeyHandler
   [-Bound]
   [-Unbound]
   [<CommonParameters>]

Get-PSReadLineKeyHandler
   [-Chord] <String[]>
   [<CommonParameters>]
```

It seems I have multiple versions installed

```PowerShell
PS Repo:\> (Get-Module -Listavailable) | Where-Object -FilterScript { $_.Name -Contains 'PSReadLine'}

    Directory: C:\program files\powershell\7\Modules

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Script     2.2.6                 PSReadLine                          Desk      {Get-PSReadLineKeyHandler, Set-PSReadLineKeyHandler, Remove-PSReadLineKeyHandler, Get-PSReadLineOption…}

    Directory: C:\Program Files\WindowsPowerShell\Modules

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Script     2.0.0                 PSReadLine                          Desk      {Get-PSReadLineKeyHandler, Set-PSReadLineKeyHandler, Remove-PSReadLineKeyHandler, Get-PSReadLineOption…}

    Directory: C:\Users\{user}\.vscode\extensions\ms-vscode.powershell-2023.1.0\modules

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Script     2.2.6                 PSReadLine                          Desk      {Get-PSReadLineKeyHandler, Set-PSReadLineKeyHandler, Remove-PSReadLineKeyHandler, Get-PSReadLineOption…}
```

### PowerShell Version

``` console
PS Repo:\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.3.2
PSEdition                      Core
GitCommitId                    7.3.2
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visual Studio Code Version

```
PS Repo:\> code --version
1.75.0
e2816fe719a4026ffa1ee0189dc89bdfdbafb164
x64
```

### Extension Version

```PowerShell
PS Repo:\> code --list-extensions --show-versions | Select-String powershell

ms-vscode.powershell@2023.1.0
ms-vscode.powershell-preview@2023.2.0
```

### Steps to Reproduce

1. Run the following command in the integrated terminal. Or dot source a script with the command.
I tried multiple commands. Always the same error. when using powershell.exe in the terminal or script.
```PowerShell
PS Repo:\> powershell.exe -Command "Get-Module"
```

2. Observe the command always puts out an error with the result of your actual command at the end

```PowerShell
Get-PSReadLineKeyHandler : A parameter cannot be found that matches parameter name 'Chord'.
At C:\Users\{user}\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\contrib\terminal\browser\media\shellIntegration.ps1:98 char:40        
+     $Handler = $(Get-PSReadLineKeyHandler -Chord $Chord | Select-Obje ...
+                                           ~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Get-PSReadLineKeyHandler], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Microsoft.PowerShell.GetKeyHandlerCommand

Get-PSReadLineKeyHandler : A parameter cannot be found that matches parameter name 'Chord'.
At C:\Users\{user}\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\contrib\terminal\browser\media\shellIntegration.ps1:98 char:40        
+     $Handler = $(Get-PSReadLineKeyHandler -Chord $Chord | Select-Obje ...
+                                           ~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Get-PSReadLineKeyHandler], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Microsoft.PowerShell.GetKeyHandlerCommand

Get-PSReadLineKeyHandler : A parameter cannot be found that matches parameter name 'Chord'.
At C:\Users\{user}\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\contrib\terminal\browser\media\shellIntegration.ps1:98 char:40        
+     $Handler = $(Get-PSReadLineKeyHandler -Chord $Chord | Select-Obje ...
+                                           ~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Get-PSReadLineKeyHandler], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Microsoft.PowerShell.GetKeyHandlerCommand

Get-PSReadLineKeyHandler : A parameter cannot be found that matches parameter name 'Chord'.
At C:\Users\{user}\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\contrib\terminal\browser\media\shellIntegration.ps1:98 char:40        
+     $Handler = $(Get-PSReadLineKeyHandler -Chord $Chord | Select-Obje ...
+                                           ~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Get-PSReadLineKeyHandler], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Microsoft.PowerShell.GetKeyHandlerCommand


ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Manifest   3.1.0.0    Microsoft.PowerShell.Management     {Add-Computer, Add-Content, Checkpoint-Computer, Clear-Content...}
Manifest   3.1.0.0    Microsoft.PowerShell.Utility        {Add-Member, Add-Type, Clear-Variable, Compare-Object...}
Script     2.0.0      PSReadLine                          {Get-PSReadLineKeyHandler, Get-PSReadLineOption, Remove-PSReadLineKeyHandler, Set-PSReadLineKeyHandler...} 

```

**Comment by DesuNinshu** [comment]
> Thanks @DesuNinshu can you try running Get-Module without -ListAvailable so we can confirm the version of PSReadLine being used?

I have that in the post already I am re-pasting it below though.

```PowerShell
PS Repo:\> (Get-Module -Listavailable) | Where-Object -FilterScript { $_.Name -Contains 'PSReadLine'}

    Directory: C:\program files\powershell\7\Modules

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Script     2.2.6                 PSReadLine                          Desk      {Get-PSReadLineKeyHandler, Set-PSReadLineKeyHandler, Remove-PSReadLineKeyHandler, Get-PSReadLineOption…}

    Directory: C:\Program Files\WindowsPowerShell\Modules

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Script     2.0.0                 PSReadLine                          Desk      {Get-PSReadLineKeyHandler, Set-PSReadLineKeyHandler, Remove-PSReadLineKeyHandler, Get-PSReadLineOption…}

    Directory: C:\Users\{user}\.vscode\extensions\ms-vscode.powershell-2023.1.0\modules

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Script     2.2.6                 PSReadLine                          Desk      {Get-PSReadLineKeyHandler, Set-PSReadLineKeyHandler, Remove-PSReadLineKeyHandler, Get-PSReadLineOption…}

```

**Comment by DesuNinshu** [comment]
I dont know if this is helpful but I output the raw clixml output from the command I will attach it.
```PowerShell
(powershell.exe -o XML -Command "Get-Module") | Out-File -FilePath .\output.xml -Force
```
[output.log](https://github.com/PowerShell/vscode-powershell/files/10703356/output.log)

**Comment by andschwa** [comment]
Is there a chance that you are starting a Windows PowerShell terminal, e.g. `powershell.exe`, and _from that terminal_ starting up VS Code by like `code.exe` or something? Somehow Windows PowerShell's PSReadLine is getting loaded, which isn't going to work, so we need to figure out the circumstances causing that to happen for you.

**Comment by DesuNinshu** [comment]
> Is there a chance that you are starting a Windows PowerShell terminal, e.g. `powershell.exe`, and _from that terminal_ starting up VS Code by like `code.exe` or something? Somehow Windows PowerShell's PSReadLine is getting loaded, which isn't going to work, so we need to figure out the circumstances causing that to happen for you.

No, the terminal that I am using is the one that is loaded by the extension with `powershell.startAutomatically` enabled.

I was able to reproduce this issue by running those commands as part of a script that was dot sourced in the PowerShell Extension terminal as well. PSReadLine is a default module that comes with PowerShell 5.1.

Additionally, the command doesn't fail. It outputs the error and the result of the command almost like it ran the command twice. It just seems like when running a script or command in the terminal that includes `powershell.exe` in it. It is loading the modules that come with 5.1 temporarily. I could be wrong though.

**Comment by DesuNinshu** [comment]
# Update:

I attempted to updated PSReadLine in Windows PowerShell (Desktop). I then had both versions 2.0.0 and 2.2.6 in the modules of windows powershell.
```PowerShell
[ADMIN]: PS C:\> Install-Module -Name PSReadLine -Scope AllUsers -Force
[ADMIN]: PS C:\> (Get-Module -Listavailable) | Where-Object -FilterScript { $_.Name -Contains 'PSReadLine'}


    Directory: C:\Program Files\WindowsPowerShell\Modules


ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     2.2.6      PSReadLine                          {Get-PSReadLineKeyHandler, Set-PSReadLineKeyHandler, Remove-PSReadLineKeyHandler, Get-PSReadLineOption...}
Script     2.0.0      PSReadLine                          {Get-PSReadLineKeyHandler, Set-PSReadLineKeyHandler, Remove-PSReadLineKeyHandler, Get-PSReadLineOption...}
```

So I tried to remove the module. By specifying the version.
```PowerShell
[ADMIN]: PS C:\> Uninstall-Module -Name PSReadLine -RequiredVersion 2.0.0
PackageManagement\Uninstall-Package : No match was found for the specified search criteria and module names 'PSReadLine'.
At C:\Program Files\WindowsPowerShell\Modules\PowerShellGet\2.2.5\PSModule.psm1:12733 char:21
+ ...        $null = PackageManagement\Uninstall-Package @PSBoundParameters
+                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Microsoft.Power...ninstallPackage:UninstallPackage) [Uninstall-Package], Exception
    + FullyQualifiedErrorId : NoMatchFound,Microsoft.PowerShell.PackageManagement.Cmdlets.UninstallPackage
```
And that didn't work so I tried removing all versions.
```PowerShell
[ADMIN]: PS C:\> Uninstall-Module -Name PSReadLine -AllVersions
[ADMIN]: PS C:\> Uninstall-Module -Name PSReadLine -AllVersions -Force
PackageManagement\Uninstall-Package : No match was found for the specified search criteria and module names 'PSReadLine'.
At C:\Program Files\WindowsPowerShell\Modules\PowerShellGet\2.2.5\PSModule.psm1:12733 char:21
+ ...        $null = PackageManagement\Uninstall-Package @PSBoundParameters
+                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Microsoft.Power...ninstallPackage:UninstallPackage) [Uninstall-Package], Exception
    + FullyQualifiedErrorId : NoMatchFound,Microsoft.PowerShell.PackageManagement.Cmdlets.UninstallPackage

[ADMIN]: PS C:\> (Get-Module -Listavailable) | Where-Object -FilterScript { $_.Name -Contains 'PSReadLine'}


    Directory: C:\Program Files\WindowsPowerShell\Modules


ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     2.0.0      PSReadLine                          {Get-PSReadLineKeyHandler, Set-PSReadLineKeyHandler, Remove-PSReadLineKeyHandler, Get-PSReadLineOption...}
```

I cannot remove the module or upgrade it. And from what I can tell the reason my issue is happening is because PowerShell Core was meant to unify powershell versions and it has the ability to read the PSModulePath whereas PowerShell Desktop can only see the modules installed in its folder and its dependency modules which are protected. 

Calling the executable file powershell.exe from a PowerShell Core terminal in vscode created by the PowerShell extension or via an integrated profile result in the same error. But, calling powershell.exe in the external version of PowerShell Core does not result in the error and does not list any modules being loaded.

I tried using Get-Command in my external terminal instead, which did return one version of the command. The one that shouldnt be returned.
```PowerShell
[ADMIN]: PS C:\> powershell.exe -Command "Get-Command Get-PSReadLineKeyHandler"

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Cmdlet          Get-PSReadLineKeyHandler                           2.0.0      PSReadLine
```

## Solution Suggestion

You could use the old 5.1 version of the command to filter for the Key/Function that is needed? I tested the function below I am not getting the error anymore.

``` PowerShell
function Set-MappedKeyHandler {
	Param ( [string[]]$Chord, [string[]]$Sequence )
	try {
		$Handler = (Get-PSReadLineKeyHandler -Chord $Chord | Select-Object -First 1)
	} catch {
		$Handler = (Get-PSReadLineKeyHandler -Bound | Where-Object -FilterScript { $_.Key -eq $Chord } | Select-Object -First 1)
	} finally {
		if ($Handler) {
			Set-PSReadLineKeyHandler -Chord $Sequence -Function $Handler.Function
		}
	} 
}

If you need more information from me let me know otherwise, I think it will be fine to close the issue or whatever you guys' think is best.

**Comment by DesuNinshu** [comment]
Repo info has been added

**Comment by DesuNinshu** [comment]
@andschwa @SydneyhSmith Should this be closed? I havent heard back from you guys I added a lot of info.

**Comment by andschwa** [comment]
Hi @DesuNinshu, the extension is [hard-coded](https://github.com/PowerShell/PowerShellEditorServices/blob/33f82b9a9665b13f86823cff3f04f7d983aae322/src/PowerShellEditorServices/Services/PowerShell/Console/PSReadLineProxy.cs#L52) to load PSReadLine from the bundled modules path (that is, the PSReadLine module shipped inside the extension itself). With the behavior you're describing, it seems as if something _else_ is loading it first. Which is bizarre since [it's the first thing we try to load](https://github.com/PowerShell/PowerShellEditorServices/blob/33f82b9a9665b13f86823cff3f04f7d983aae322/src/PowerShellEditorServices/Services/PowerShell/Host/PsesInternalHost.cs#L1163). Installing/uninstalling PSReadLine module should not affect what the Extension Terminal is using.

So...I took a look at what you've shared again and most of your examples are not _in_ the Extension Terminal. For instance, this error you shared:

> Get-PSReadLineKeyHandler : A parameter cannot be found that matches parameter name 'Chord'.
At C:\Users\{user}\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\contrib\terminal\browser\media\shellIntegration.ps1:98 char:40

Shows that the built-in VS Code shell integration script is being run, which means you're in a regular VS Code Integrated Terminal, completely outside the extension.

Your suggestion is probably a good one to improve the PowerShell 5.1 support of the integrated terminal script, but that's going to need to be done upstream with [VS Code](https://github.com/microsoft/vscode). I have worked with them to improve their script for better PowerShell support, and we can try to get this fixed, but I'm marking this particular issue as resolved because it's not actually related to the PowerShell Extension.
